modules/layers/src/path-layer/path-tesselator.js

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2015 - 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+import Tesselator from './tesselator';
+import {fp64 as fp64Module} from 'luma.gl';
+const {fp64LowPart} = fp64Module;
+
+// This class is set up to allow querying one attribute at a time
+// the way the AttributeManager expects it
+export default class PathTesselator extends Tesselator {
+  constructor({data, getGeometry, fp64}) {
+    super({
+      data,
+      getGeometry,
+      fp64,
+      attributes: {
+        startPositions: {size: 3},
+        endPositions: {size: 3},
+        leftDeltas: {size: 3},
+        rightDeltas: {size: 3},
+        startEndPositions64XyLow: {size: 4, fp64: true}
+      }
+    });
+  }
+
+  /* Getters */
+  startPositions() {
+    return this.attributes.startPositions;
+  }
+
+  endPositions() {
+    return this.attributes.endPositions;
+  }
+
+  startEndPositions64XyLow() {
+    return this.attributes.startEndPositions64XyLow;
+  }
+
+  leftDeltas() {
+    return this.attributes.leftDeltas;
+  }
+
+  rightDeltas() {
+    return this.attributes.rightDeltas;
+  }
+
+  strokeWidths({target, getWidth}) {
+    return this._updateAttribute({
+      target,
+      size: 1,
+      getValue: object => [getWidth(object)]
+    });
+  }
+
+  dashArrays({target, getDashArray}) {
+    return this._updateAttribute({target, size: 2, getValue: getDashArray});
+  }
+
+  colors({target, getColor}) {
+    return this._updateAttribute({
+      target,
+      size: 4,
+      getValue: object => {
+        const color = getColor(object);
+        if (isNaN(color[3])) {
+          color[3] = 255;
+        }
+        return color;
+      }
+    });
+  }
+
+  pickingColors({target, getPickingColor}) {
+    return this._updateAttribute({
+      target,
+      size: 3,
+      getValue: (object, index) => getPickingColor(index)
+    });
+  }
+
+  /* Utilities */
+  getPathLength(path) {
```

Comment:
The utilities are really private methods used by the "virtual tesselator functions", and as such I would place them last in the class.